### PR TITLE
Make View::shmem_size() device-callable

### LINKED ISSUE #4925: Make View::shmem_size() device callable
Currently, View::shmem_size() does not have KOKKOS_*_FUNCTION markings, I'd find it useful to be able to call it inside a kernel. Could the markings be added?

### PR DESCRIPTION
**Pull request by ldh4**
To resolve #4925

### CODE REVIEW

**Comment by ldh4** [comment]
Please retest this.
Changes made in this commit to add `KOKKOS_INLINE_FUNCTION` are only relevant to `Cuda` and `HIP` backends.
I am not sure why a performance test on `SYCL` backend is affected by this.

**Comment by ajpowelsnl** [comment]
Check for spurious failures.

**Review by cz4rs — APPROVED** [review]
Looks good to me.

File: core/unit_test/TestViewAPI_e.hpp [hunk 1]
```diff
@@ -268,6 +268,43 @@ TEST(TEST_CATEGORY, view_allocation_large_rank) {
       Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, v_single);
   ASSERT_EQ(result(0, 0, 0, 0, 0, 0, 0, 0), 42);
 }
+
+template <typename ExecSpace, typename ViewType>
+struct TestViewShmemSizeOnDevice {
+  using ViewTestType = Kokkos::View<size_t*, ExecSpace>;
```
Comment (dalg24) [inline core/unit_test/TestViewAPI_e.hpp:274]: Prefer a rank-0 view

**Review by dalg24 — APPROVED** [review]

File: core/unit_test/TestViewAPI_e.hpp (hunk 1 above)
Comment (ldh4) [reply core/unit_test/TestViewAPI_e.hpp]: Ok, replaced it.

**Pull request merged**